Create a Raw Data Table

GabrielMissael opened the pull request:
## Changes description.

This PR adds three new tables in the database to store raw data collected during experiments. The new tables are shown in the following figure:

<p align="center">
  <img alt="Database structure" src="https://i.imgur.com/lr7NTSj.png" width="80%">
</p>

The specific changes in the code are in 6 different files, described below:

1. `aepsych/database/tables.py`: Add three new tables structures to the database:
    1. Raw data table: fact table for raw data collected during experiments.
    2. Param table: dimension table for parameters used in the experiment.
    3. Outcome table: dimension table for outcomes of the experiment.

2. `aepsych/database/db.py`: Integrates these new tables into the database. Methods to save data, retrieve data, and update the from the tables are included. Also, there is a new method `generate_experiment_table`, which generates a new table with raw data for a given experiment id. This table is created within the database and intended to make it easy to access an experiment's data. An example of this table structure:

| iteration_id 	| theta1 	| theta2 	| outcome_0 	| outcome_1 	|                  timestamp 	|
|-------------:	|-------:	|-------:	|----------:	|----------:	|---------------------------:	|
|            1 	|    1.0 	|    0.0 	|       0.0 	|     -10.0 	| 2022-10-17 10:52:23.757922 	|
|            2 	|    3.0 	|    2.0 	|       1.0 	|      -7.5 	| 2022-10-17 10:52:23.857462 	|
|            3 	|    5.0 	|    4.0 	|       0.0 	|      -5.0 	| 2022-10-17 10:52:23.949986 	|
|            4 	|    7.0 	|    6.0 	|       1.0 	|      -2.5 	| 2022-10-17 10:52:24.041166 	|

4. `aepsych/server/server.py`: Saves experiment results when the `tell()` method is used in the server class, and it's not a replay.
5. `aepsych/tests/test_db.py`: Add tests for the new tables.
6. `aepsych/tests/test_integration`: These tests check that the server can handle different experiments (multi/single stimuli, multi/single outcome). They ensure that the data is correctly stored in the database tables (raw, param, and outcome). It also checks that the experiment table is correctly populated (generate_experiment_table method).

The PR https://github.com/facebookresearch/aepsych/pull/184 adds the descriptions of the DB changes to the `For Developers/Data Overview section`.

## Associated Issue
Fixes #34

## To-Do

- [x] Implement tests for new table.
- [x] Create `--update` method to apply these changes to old databases.
    - This goes in a new PR. Associated branch: https://github.com/GabrielMissael/aepsych/tree/update_database
- [x] The outcome is not necessarily binary!
- [x] Allow multiple stimuli per experiment
- [x] Allow multiple outcomes per experiment

GabrielMissael commented:
Hi, I'm running the tests locally. I just fixed the errors that popped out in this PR. Now, I only get two errors, but I'm not sure how to fix them.

```
======================================================================
FAIL: test_serialization_2d (tests.models.test_pairwise_probit.PairwiseProbitModelServerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "~/aepsych/tests/models/test_pairwise_probit.py", line 683, in test_serialization_2d
    server2.tell(config=next_config, outcome=next_y)
AttributeError: 'AEPsychServer' object has no attribute 'db'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/aepsych/tests/models/test_pairwise_probit.py", line 685, in test_serialization_2d
    self.fail()
AssertionError: None
----------------------------------------------------------------------
```

And the same error for `test_serialization_1d`. This is the `try` block that is not working:

```python
        import dill

        # just make sure it works
        try:
            s = dill.dumps(server)
            server2 = dill.loads(s)
            next_config = server2.ask()
            next_pair = np.c_[next_config["x"], next_config["y"]].T
            next_y = bernoulli.rvs(f_pairwise(f_2d, next_pair))
            server2.tell(config=next_config, outcome=next_y)
        except Exception:
            self.fail()
```

So perhaps it's an error with `dill`? Because `server2` should be equivalent to `server`, which does have a `db` attribute. I'm running this in WSL2, with Ubuntu 20.04.4, and with the specified packages in requirements and dev_requirements.

Thanks in advance! 😄

mshvartsman commented:
When pickling the server (mostly for crash dumps) we don't include the db or socket objects, since those reference things we can't unpickle (because the socket is transient and the db is disk-backed). This happens here https://github.com/facebookresearch/aepsych/blob/main/aepsych/server/server.py#L763. Previously this wasn't a problem in this test because the db wasn't used in `tell`. Ultimately the goal of the test is to just make sure that the pickled server has the information we need to debug crash dumps (which is mostly the model and data). You can replace the `ask` and `tell` with some asserts that the unpickled server has data, strategy, and models. Here's an overkill set of asserts: 

1. Checking strats: server2._strats is a list of instances of the same types as server._strats. 
2. Checking models: if any strat in server._strats has a model attribute, the same element in server2._strats also has that attribute, and that the models are the same types. 
3. Checking data: for any strat in server._strats, `strat.x` and `strat.y` are the same as those items in `server2`. 

In practice I imagine if one of these succeeds all will succeed and if one fails all will fail -- but writing them all out might get you nice exposure to the structure of these objects :).

crasanders reviewed (commented):
This looks good, but I wonder if we could have stronger tests. Specifically, I think we should have a test for `generate_experiment_table` and at least one integration test. The integration test could configure a server, run some asks and tells, and then verify that the resulting tables contain the correct data.

GabrielMissael commented on tests/test_integration.py:
```diff
@@ -148,11 +148,87 @@ def test_single_stimuli_single_outcome(self):
+        i = 0
+        while not self.s.strat.finished and i<7:
+            #! TODO: Ask request should be supported!
+            # self.s.unversioned_handler(self.ask_request)
```
The ask-request is not working in this case (single stimuli with multiple outcomes). I'm not sure if it's because multiple outcomes are still not supported or if I need to change something in the config file `tests/configs/singleStimuli.ini` 🤔

Here is the full error message of this test case when this code line (ask-request) is uncommented:

```
======================================================================
ERROR: test_single_stimuli_multi_outcome (tests.test_integration.IntegrationTestCase)
Test a single-stimulus experiment with a multiple outcomes.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/missa/projects/aepsych/tests/test_integration.py", line 167, in test_single_stimuli_multi_outcome
    self.s.unversioned_handler(self.ask_request)
  File "/home/missa/projects/aepsych/aepsych/server/server.py", line 467, in unversioned_handler
    ret_val = message_map[type](request)
  File "/home/missa/projects/aepsych/aepsych/server/server.py", line 506, in handle_ask
    new_config = self.ask()
  File "/home/missa/projects/aepsych/aepsych/server/server.py", line 721, in ask
    next_x = self.strat.gen()[0]
  File "/home/missa/projects/aepsych/aepsych/strategy.py", line 460, in gen
    return self._strat.gen(num_points=num_points, **kwargs)
  File "/home/missa/projects/aepsych/aepsych/strategy.py", line 33, in wrapper
    self.fit()
  File "/home/missa/projects/aepsych/aepsych/strategy.py", line 315, in fit
    self.model.fit(self.x, self.y)
  File "/home/missa/projects/aepsych/aepsych/models/gp_classification.py", line 231, in fit
    self._fit_mll(mll, **kwargs)
  File "/home/missa/projects/aepsych/aepsych/models/base.py", line 426, in _fit_mll
    fit_gpytorch_mll(mll, optimizer_kwargs=optimizer_kwargs, **kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/fit.py", line 114, in fit_gpytorch_mll
    return dispatcher(
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/utils/dispatcher.py", line 88, in __call__
    return func(*args, **kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/fit.py", line 240, in _fit_fallback
    mll, _ = optimizer(mll, **optimizer_kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/optim/fit.py", line 142, in fit_gpytorch_scipy
    res = minimize(
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_minimize.py", line 699, in minimize
    res = _minimize_lbfgsb(fun, x0, args, jac, bounds,
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_lbfgsb_py.py", line 306, in _minimize_lbfgsb
    sf = _prepare_scalar_function(fun, x0, jac=jac, args=args, epsilon=eps,
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_optimize.py", line 263, in _prepare_scalar_function
    sf = ScalarFunction(fun, x0, args, grad, hess,
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 158, in __init__
    self._update_fun()
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 251, in _update_fun
    self._update_fun_impl()
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 155, in update_fun
    self.f = fun_wrapped(self.x)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_differentiable_functions.py", line 137, in fun_wrapped
    fx = fun(np.copy(x), *args)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_optimize.py", line 76, in __call__
    self._compute_if_needed(x, *args)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/scipy/optimize/_optimize.py", line 70, in _compute_if_needed
    fg = self.fun(x, *args)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/optim/utils.py", line 239, in _scipy_objective_and_grad
    return _handle_numerical_errors(error=e, x=x)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/optim/utils.py", line 268, in _handle_numerical_errors
    raise error  # pragma: nocover
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/botorch/optim/utils.py", line 237, in _scipy_objective_and_grad
    loss = -mll(*args).sum()
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/module.py", line 30, in __call__
    outputs = self.forward(*inputs, **kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/mlls/variational_elbo.py", line 77, in forward
    return super().forward(variational_dist_f, target, **kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/mlls/_approximate_mll.py", line 58, in forward
    log_likelihood = self._log_likelihood_term(approximate_dist_f, target, **kwargs).div(num_batch)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/mlls/variational_elbo.py", line 61, in _log_likelihood_term
    return self.likelihood.expected_log_prob(target, variational_dist_f, **kwargs).sum(-1)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/likelihoods/bernoulli_likelihood.py", line 53, in expected_log_prob
    log_prob = self.quadrature(log_prob_lambda, function_dist)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/utils/quadrature.py", line 81, in forward
    log_probs = func(shifted_locs)
  File "/home/missa/projects/aepsych/mlh_env_ubuntu/lib/python3.8/site-packages/gpytorch/likelihoods/bernoulli_likelihood.py", line 52, in <lambda>
    log_prob_lambda = lambda function_samples: log_normal_cdf(function_samples.mul(observations))
RuntimeError: The size of tensor a (3) must match the size of tensor b (6) at non-singleton dimension 2

----------------------------------------------------------------------
```

GabrielMissael commented on tests/test_integration.py:
```diff
@@ -18,12 +18,30 @@
+all_tests = ['singleStimuli_singleOutcome',
+             'multiStimuli_singleOutcome', 'multiStimuli_multiOutcome']
+#! This should be supported! Uncoment when ready
+# all_tests.append('singleStimuli_multiOutcome')
```
This test case did not pass. I commented this test case because I think multi-outcome it's still not suported🤔

mshvartsman commented on tests/test_integration.py:
```diff
@@ -0,0 +1,171 @@
+outcomes = {'singleOutcome':[1, -1, 0.1, 0, -0.1, 0, 0],
+            'multiOutcome':[[1, 0], [-1, 0], [0.1, 0], [0, 0], [-0.1, 0], [0, 0], [0, 0]]}
+
+all_tests = ['singleStimuli_singleOutcome',
```
Minor nit -- if you just make the tests tuples of `stimuli, outcome` objects (which you can generate from an `itertools.product(params, outcomes)` then you don't have to construct these strings and then parse them later.

mshvartsman commented on tests/test_integration.py:
```diff
@@ -0,0 +1,171 @@
+}}
+
+outcomes = {'singleOutcome':[1, -1, 0.1, 0, -0.1, 0, 0],
+            'multiOutcome':[[1, 0], [-1, 0], [0.1, 0], [0, 0], [-0.1, 0], [0, 0], [0, 0]]}
```
I think this needs an additional level of nesting on these lists for the hacky way multi-outcome works today. So `[[1], [0]], [[-1], [0]]]...`. Sorry it's so hacky for now.

GabrielMissael replied:
Thank you for the clarification! All tests are working now ✨

mshvartsman commented on aepsych/server/server.py:
```diff
@@ -297,6 +298,45 @@ def _flatten(x):
             )
         return out
 
+    def generate_experiment_table(self, experiment_id=None, table_name = 'experiment_table',
```
With the addition of this method, there's two canonical ways of getting data out of a db. One of them is with this method, and the other is with `get_dataframe_from_replay` elsewhere in this file (which uses the replay or strat tables depending on what's available). I think that you should add a `DeprecationWarning` to the other method which recommends that users use this method instead. Other than that this PR looks pretty good!

GabrielMissael replied:
Thank you! ✨

I just added the deprecation warning 👀

GabrielMissael commented:
Hey there! 😄 
I have updated the code format using UFMT as recommended. Also, I added a docstring to the `generate_experiment_table` method. I'm not sure to add docstrings to all the other methods because I think they are not public methods ([PEP 8](http://www.python.org/dev/peps/pep-0008/))

mshvartsman commented:
FYI, linter errors look like a recent breaking change in mypy related to PEP 484 (https://mypy-lang.blogspot.com/2022/11/mypy-0990-released.html). @GabrielMissael or @Pravallika-Myneni if you have time to put up a fix for that that'd be cool, otherwise @crasanders or me will get to it later this week.

GabrielMissael commented:
Hi @mshvartsman!

I'll take a look at it to put up a fix!

crasanders reviewed (changes requested):
This is close. You just need to add newlines to the end of the config files and remove the unused imports from your integration test.

GabrielMissael reviewed (commented):
Done! ✨

mshvartsman commented on tests/test_integration.py:
```diff
@@ -0,0 +1,223 @@
+        exp_id = self.s.db.get_master_records()[0].experiment_id
+
+        # Get tables data
+        raw_data = self.s.db.get_raw_for(exp_id)
```
I believe all three of these lines (207-209) have variables that are assigned but never used (our internal linter is catching this, not sure why OSS isn't, cc @crasanders in case something in github actions can be improved). What is the intent? If it is just to ensure they work and don't throw, you can do: 
```
try: 
    _ = self.s.db.get_raw_for(exp_id)
except Exception as e: # ideally a more specific exception though 
    self.fail(f"self.s.db.get_raw_for failed with exception {e}") #
or whatever error message you want to fail with. 
```
(via https://stackoverflow.com/questions/4319825/python-unittest-opposite-of-assertraises)

Or if you meant additional checks here, you should add them.

GabrielMissael replied:
Oh sorry! I considered comparing them with the values saved in the experiment table. Still, in the end, I checked the values in the experiment table with the original values from the dictionaries!

I'll remove those variables! ✅

mshvartsman commented on tests/test_integration.py:
```diff
@@ -0,0 +1,223 @@
+
+import aepsych.server as server
+import aepsych.utils_logging as utils_logging
+import torch
```
This isn't used right? Can remove.

GabrielMissael replied:
Yes! You're right!
Thank you ✨
